### PR TITLE
Pretty print async fn sugar in opaques and trait bounds

### PR DESCRIPTION
**Pull request by compiler-errors**
sudo r? fmease

### CODE REVIEW

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 1]
```diff
@@ -1042,88 +1040,66 @@ pub trait PrettyPrinter<'tcx>: Printer<'tcx> + fmt::Write {
...
             write!(self, "{}", if first { "" } else { " + " })?;
             write!(self, "{}", if paren_needed { "(" } else { "" })?;
 
-            self.wrap_binder(&fn_once_trait_ref, |trait_ref, cx| {
-                define_scoped_cx!(cx);
-                // Get the (single) generic ty (the args) of this FnOnce trait ref.
-                let generics = tcx.generics_of(trait_ref.def_id);
-                let own_args = generics.own_args_no_defaults(tcx, trait_ref.args);
-
-                match (entry.return_ty, own_args[0].expect_ty()) {
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1055]: I have noooooooOOoOooooooo idea wtf was going on here lmao

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 2]
```diff
@@ -1230,51 +1206,48 @@ pub trait PrettyPrinter<'tcx>: Printer<'tcx> + fmt::Write {
     /// traits map or fn_traits map, depending on if the trait is in the Fn* family of traits.
     fn insert_trait_and_projection(
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1207]: I'm pretty disappointed with how sensitive to bound var renumbering this is... but 🤷

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 3]
```diff
@@ -1042,88 +1040,66 @@ pub trait PrettyPrinter<'tcx>: Printer<'tcx> + fmt::Write {
...
-                        for (idx, ty) in arg_tys.tuple_fields().iter().enumerate() {
-                            if idx > 0 {
-                                p!(", ");
-                            }
-                            p!(print(ty));
-                        }
+            if let Some(return_ty) = entry.return_ty {
+                self.wrap_binder(&bound_args, |args, cx| {
+                    define_scoped_cx!(cx);
+                    p!(write("{}", tcx.item_name(trait_def_id)));
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1056]: we could very easily start making this print like `async Fn` in the future.

**Review by fmease — APPROVED** [review]
Thanks, one small suggestion

**Comment by fmease** [comment]
Then r=me @bors rollup

File: compiler/rustc_middle/src/ty/print/pretty.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_middle/src/ty/print/pretty.rs:1055]: Unfortunately, something about this part was load-bearing, cf. #133597

Comment (jieyouxu) [reply compiler/rustc_middle/src/ty/print/pretty.rs:1055]: Unfortunately, something about this part was load-bearing, cf. #133597

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 4]
```diff
@@ -1042,88 +1038,66 @@ pub trait PrettyPrinter<'tcx>: Printer<'tcx> + fmt::Write {
...
-            })?;
+                    first = false;
+                    Ok(())
+                })?;
+            } else {
+                // Otherwise, render this like a regular trait.
+                traits.insert(
+                    bound_args.map_bound(|args| ty::TraitPredicate {
+                        polarity: ty::PredicatePolarity::Positive,
+                        trait_ref: ty::TraitRef::new(tcx, trait_def_id, [Ty::new_tup(tcx, args)]),
```
Comment (fmease) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1080]: A missing the self ty if applicable

**Pull request merged**